crates/goose-mcp/src/todo/mod.rs

```diff
@@ -0,0 +1,134 @@
+use std::future::Future;
+use std::pin::Pin;
+
+use mcp_core::handler::{PromptError, ResourceError};
+use mcp_core::protocol::ServerCapabilities;
+use mcp_server::router::CapabilitiesBuilder;
+use mcp_server::Router;
+use rmcp::model::{Content, ErrorCode, ErrorData, JsonRpcMessage, Prompt, Resource, Tool};
+use serde_json::Value;
+use tokio::sync::mpsc;
+
+// Import the TODO tools
+use goose::agents::todo_tools::{todo_read_tool, todo_write_tool};
+
+/// A lightweight router that provides TODO task management capabilities.
+///
+/// This extension acts as a metadata provider - it exposes the TODO tools
+/// and provides instructions, but the actual execution remains in the agent
+/// for session storage access.
+#[derive(Clone)]
+pub struct TodoRouter {
+    tools: Vec<Tool>,
+    instructions: String,
+}
+
+impl Default for TodoRouter {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl TodoRouter {
+    pub fn new() -> Self {
+        let instructions = r#"The todo extension provides persistent task management throughout your session.
+
+These tools help you track multi-step work, maintain context between interactions, and ensure systematic task completion.
+
+## Task Management Guidelines
+
+**Required Usage:**
+- Use `todo__read` and `todo__write` for any task with 2+ steps, multiple files/components, or uncertain scope
+- Skipping these tools when needed is considered an error
+
+**Workflow:**
+1. Start: Always `todo__read` first, then `todo__write` a brief checklist using Markdown checkboxes
+2. During: After each major action, reread via `todo__read`, then update via `todo__write` - mark completed items, add new discoveries, note blockers
+3. Finish: Ensure every item is checked, or clearly list what remains
+
+**Critical:** `todo__write` replaces the entire list. Always read before writing - not doing so is an error.
+
+**Best Practices:**
+- Keep items short, specific, and action-oriented
+- Use nested checkboxes for subtasks
+- Include context about blockers or dependencies
+
+Example format:
+```markdown
+- [x] Analyze request fully
```

Comment:
I think if we are moving the prompt here we need to make it clear similar to the old system.md one how to do it, including if we want it to use subagents at times (otherwise I found it will not) - https://github.com/block/goose/pull/4421/files#diff-507a51de1f94529d2674a027f5bf6e5eeed83ccfe9cbbe6cea674b75aef65ee8 like that